diff --git a/browser/devtools/debugger/test/browser.ini b/browser/devtools/debugger/test/browser.ini
index eeba980ae5b8..6527fc516046 100644
--- a/browser/devtools/debugger/test/browser.ini
+++ b/browser/devtools/debugger/test/browser.ini
@@ -103,71 +103,45 @@ support-files =
 [browser_dbg_aaa_run_first_leaktest.js]
 skip-if = e10s && debug
 [browser_dbg_addonactor.js]
-skip-if = e10s && debug
 [browser_dbg_addon-sources.js]
-skip-if = e10s && debug
 [browser_dbg_addon-modules.js]
 skip-if = e10s # TODO
 [browser_dbg_addon-modules-unpacked.js]
 skip-if = e10s # TODO
 [browser_dbg_addon-panels.js]
-skip-if = e10s && debug
 [browser_dbg_addon-console.js]
 skip-if = e10s && debug || os == 'win' # bug 1005274
 [browser_dbg_auto-pretty-print-01.js]
-skip-if = e10s && debug
 [browser_dbg_auto-pretty-print-02.js]
-skip-if = e10s && debug
 [browser_dbg_bfcache.js]
 skip-if = e10s || true # bug 1113935
 [browser_dbg_blackboxing-01.js]
-skip-if = e10s && debug
 [browser_dbg_blackboxing-02.js]
-skip-if = e10s && debug
 [browser_dbg_blackboxing-03.js]
-skip-if = e10s && debug
 [browser_dbg_blackboxing-04.js]
-skip-if = e10s && debug
 [browser_dbg_blackboxing-05.js]
-skip-if = e10s && debug
 [browser_dbg_blackboxing-06.js]
-skip-if = e10s && debug
 [browser_dbg_breadcrumbs-access.js]
-skip-if = e10s && debug
 [browser_dbg_break-on-dom-01.js]
-skip-if = e10s && debug
 [browser_dbg_break-on-dom-02.js]
-skip-if = e10s && debug
 [browser_dbg_break-on-dom-03.js]
-skip-if = e10s && debug
 [browser_dbg_break-on-dom-04.js]
-skip-if = e10s && debug
 [browser_dbg_break-on-dom-05.js]
-skip-if = e10s && debug
 [browser_dbg_break-on-dom-06.js]
-skip-if = e10s && debug
 [browser_dbg_break-on-dom-07.js]
-skip-if = e10s && debug
 [browser_dbg_break-on-dom-08.js]
-skip-if = e10s && debug
 [browser_dbg_break-on-dom-event-01.js]
 skip-if = e10s || os == "mac" || e10s # Bug 895426
 [browser_dbg_break-on-dom-event-02.js]
 skip-if = e10s # TODO
 [browser_dbg_breakpoints-actual-location.js]
-skip-if = e10s && debug
 [browser_dbg_breakpoints-actual-location2.js]
-skip-if = e10s && debug
 [browser_dbg_breakpoints-break-on-last-line-of-script-on-reload.js]
 skip-if = e10s # Bug 1093535
 [browser_dbg_breakpoints-button-01.js]
-skip-if = e10s && debug
 [browser_dbg_breakpoints-button-02.js]
-skip-if = e10s && debug
 [browser_dbg_breakpoints-contextmenu-add.js]
-skip-if = e10s && debug
 [browser_dbg_breakpoints-contextmenu.js]
-skip-if = e10s && debug
 [browser_dbg_breakpoints-disabled-reload.js]
 skip-if = e10s # Bug 1093535
 [browser_dbg_breakpoints-editor.js]
