diff --git a/netwerk/base/nsURLHelper.h b/netwerk/base/nsURLHelper.h
index b4a897382b41..94ee2b184190 100644
--- a/netwerk/base/nsURLHelper.h
+++ b/netwerk/base/nsURLHelper.h
@@ -229,7 +229,7 @@ bool net_IsValidIPv6Addr(const char *addr, int32_t addrLen);
 
 
 /**
- * Returns the max length of a URL. The default is 2083.
+ * Returns the max length of a URL. The default is 1048576 (1 MB).
  * Can be changed by pref "network.standard-url.max-length"
  */
 int32_t net_GetURLMaxLength();
