diff --git a/build/Makefile.in b/build/Makefile.in
index df27cd042bf4..25f0abf80af4 100644
--- a/build/Makefile.in
+++ b/build/Makefile.in
@@ -61,13 +61,6 @@ LLDBINIT_FINAL_TARGET_FILES := $(DEPTH)/.lldbinit
 LLDBINIT_FINAL_TARGET_DEST = $(FINAL_TARGET)
 INSTALL_TARGETS += LLDBINIT_FINAL_TARGET
 
-ifeq (1_1,$(MOZ_ASAN)_$(CLANG_CL))
-# Install the clang-cl runtime library for ASAN next to the binaries we produce.
-CLANG_RT_ASAN_FILES := $(MOZ_CLANG_RT_ASAN_LIB_PATH)
-CLANG_RT_ASAN_DEST = $(FINAL_TARGET)
-INSTALL_TARGETS += CLANG_RT_ASAN
-endif
-
 ifdef MOZTTDIR
 # Install the Firefox OS fonts.
 include $(MOZTTDIR)/fonts.mk
diff --git a/build/moz.build b/build/moz.build
index 75ac68a9d016..d30a35c1a059 100644
--- a/build/moz.build
+++ b/build/moz.build
@@ -63,3 +63,7 @@ if CONFIG['MOZ_DMD']:
 # Put a useful .gdbinit in the bin directory, to be picked up automatically
 # by GDB when we debug executables there.
 FINAL_TARGET_FILES += [TOPSRCDIR + '/.gdbinit']
+
+# Install the clang-cl runtime library for ASAN next to the binaries we produce.
+if CONFIG['MOZ_ASAN'] and CONFIG['CLANG_CL']:
+    FINAL_TARGET_FILES += [CONFIG['MOZ_CLANG_RT_ASAN_LIB_PATH']]
