diff --git a/js/src/tests/browser.js b/js/src/tests/browser.js
index 3b934d5177901..3cc54a4eb9da3 100644
--- a/js/src/tests/browser.js
+++ b/js/src/tests/browser.js
@@ -332,6 +332,10 @@ function jsTestDriverBrowserInit()
     {
       properties.version = '1.8';
     }
+    else if (properties.test.match(/^ecma_6\/Class/))
+    {
+      properties.version = '1.8';
+    }
   }
 
   // default to language=type;text/javascript. required for
